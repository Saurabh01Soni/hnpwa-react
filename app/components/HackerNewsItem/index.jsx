import React from 'react';
import PropTypes from 'prop-types';

import HackerNewsComment from 'containers/HackerNewsComment';
import HackerNewsListItem from 'components/HackerNewsListItem';

import './HackerNewsItem.scss';

const propTypes = {
  comments: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  item: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  itemId: PropTypes.number,
  onItemFetch: PropTypes.func,
};

const defaultProps = {
  comments: null,
  item: null,
  itemId: 0,
  onItemFetch() {},
};

class HackerNewsItem extends React.Component {
  componentWillMount() {
    this.props.onItemFetch(this.props.itemId);
  }

  render() {
    const { comments, item } = this.props;
    if (!comments || !item) return null;

    return (
      <div className="HackerNewsItem">
        <div className="HackerNewsItem__content">
          <HackerNewsListItem {...item.toJS()} />
        </div>
        {comments.map(id => <HackerNewsComment commentId={id} key={id} />)}
      </div>
    );
  }
}

HackerNewsItem.propTypes = propTypes;
HackerNewsItem.defaultProps = defaultProps;

export default HackerNewsItem;
