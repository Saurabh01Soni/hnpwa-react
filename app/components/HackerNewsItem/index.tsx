import { List, Map } from 'immutable';
import * as React from 'react';
import PropTypes from 'prop-types';

import HackerNewsComment from 'containers/HackerNewsComment';
import HackerNewsListItem from 'components/HackerNewsListItem';
import LoadingIndicator from 'components/LoadingIndicator';

import './HackerNewsItem.scss';

interface IHackerNewsItemProps {
  comments?: List<number> | null;
  done?: () => void;
  item?: Map<string, any> | null;
  itemId?: number;
  onItemFetch?: (arg: number) => Promise<any>;
}

class HackerNewsItem extends React.Component<IHackerNewsItemProps> {
  public static defaultProps = {
    comments: null,
    done() {},
    item: null,
    itemId: 0,
    onItemFetch(arg: number) { return Promise.resolve(); },
  };

  public componentWillMount() {
    const { done, item, itemId } = this.props;

    if (!item) {
      this.props.onItemFetch(itemId).then(done, done);
    }
  }

  public render() {
    const { comments, item } = this.props;
    if (!comments || !item) {
      return (
        <LoadingIndicator
          active
          style={{ position: 'absolute', left: 'calc(50% - 24px)', marginTop: '30px' }}
        />
      );
    }

    return (
      <div className="HackerNewsItem">
        <div className="HackerNewsItem__content">
          <HackerNewsListItem {...item.toJS()} />
        </div>
        {comments.map(id => <HackerNewsComment commentId={id} key={id} />).toArray()}
      </div>
    );
  }
}

export default HackerNewsItem;
