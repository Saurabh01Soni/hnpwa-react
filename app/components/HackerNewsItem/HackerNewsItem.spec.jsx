import Immutable from 'immutable';
import React from 'react';
import { shallow } from 'enzyme';

import HackerNewsItem from './';

describe('<HackerNewsItem />', () => {
  it('should match snapshot when render default', () => {
    const wrapper = shallow(<HackerNewsItem />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render <HackerNewsListItem /> and <HackerNewsCommentContainer />', () => {
    const item = Immutable.Map({
      commentsCount: 14,
      id: 1,
      timeAgo: '6 years ago',
      title: 'React with TypeScript',
      points: 14,
      user: 'taehwanno',
      url: 'https://github.com/taehwanno',
    });
    const comments = Immutable.List([1, 2, 3, 4]);
    const wrapper = shallow(<HackerNewsItem comments={comments} item={item} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should calls props.onItemFetch when componentWillMount is called', () => {
    const itemId = 1;
    const onItemFetch = jest.fn();
    shallow(<HackerNewsItem itemId={itemId} onItemFetch={onItemFetch} />);
    expect(onItemFetch.mock.calls.length).toBe(1);
    expect(onItemFetch.mock.calls[0]).toEqual([itemId]);
  });
});
