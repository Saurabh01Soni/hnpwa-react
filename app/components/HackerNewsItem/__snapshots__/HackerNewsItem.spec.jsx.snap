// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HackerNewsItem /> should match snapshot when render default 1`] = `null`;

exports[`<HackerNewsItem /> should render <HackerNewsListItem /> and <HackerNewsCommentContainer /> 1`] = `
<div
  className="HackerNewsItem"
>
  <div
    className="HackerNewsItem__content"
  >
    <HackerNewsListItem
      commentsCount={14}
      id={1}
      points={14}
      timeAgo="6 years ago"
      title="React with TypeScript"
      url="https://github.com/taehwanno"
      user="taehwanno"
    />
  </div>
  <Connect(HackerNewsCommentInner)
    commentId={1}
    key="1"
  />
  <Connect(HackerNewsCommentInner)
    commentId={2}
    key="2"
  />
  <Connect(HackerNewsCommentInner)
    commentId={3}
    key="3"
  />
  <Connect(HackerNewsCommentInner)
    commentId={4}
    key="4"
  />
</div>
`;
